src/contract/nodes.rs

```diff
@@ -270,12 +271,29 @@ impl ConsensusCommit for Transition {
     type Commitment = NodeId;
 }
 
-impl CommitEncodeWithStrategy for Extension {
-    type Strategy = commit_strategy::UsingStrict;
+impl CommitEncode for Extension {
```

Comment:
By doing a manual encoding we are creating a change of diverging strict and commit encodings in the future. The method should continue to use strict encoding, but on a copy of the data, after calling `conceal_all`

Reply:
> If it was never implemented than rust compiler would produce a error – since we are calling this method from the code

In reference to the above comment, the compiler does not produce any error in that regard because, in fact, that code is not being called. This can be verified by simply commenting out those line of codes (shown below) and recompiling the library. You will notice that the library will compile with no error.

```rust
/*impl CommitEncodeWithStrategy for Void {
    type Strategy = commit_strategy::UsingConceal;
}*/
```

Reply:
Not sure why the manual implementation of the trait will change the fact that the code is not called from inside rgb-core. It is not called since in our tests we do not use `void` type of data AFAIR (and it is not used in RGB20 either).

Reply:
> Not sure why the manual implementation of the trait will change the fact that the code is not called from inside rgb-core. It is not called since in our tests we do not use `void` type of data AFAIR (and it is not used in RGB20 either).

OK, fair enough. Then, as an alternative to prove my point, add the following test function to the (original) `rgb-core/src/contract/data.rs` file, and try to compile the library.

```rust
    #[test]
    fn test_void_encoding() {
        let v = Void;

        let mut e = vec![];
        let r = v.commit_encode(&mut e);

        assert_eq!(r, 0);
    }
```

You should then get the following error:

```shell
error[E0599]: no method named `commit_encode` found for struct `contract::data::Void` in the current scope
   --> src/contract/data.rs:749:19
    |
40  | pub struct Void;
    | ---------------- method `commit_encode` not found for this
...
749 |         let r = v.commit_encode(&mut e);
    |                   ^^^^^^^^^^^^^ method not found in `contract::data::Void`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following trait defines an item `commit_encode`, perhaps you need to implement it:
            candidate #1: `lnpbp::client_side_validation::CommitEncode`
```

Then, you can go a step further and make the following changes to "force" that the `CommitEncode` trait be applied to type `Void` (via `CommitEncodeWithStrategy`).

```rust
+#[derive(Clone, Debug, PartialOrd, Ord, PartialEq, Eq, AsAny)]
+pub struct ConfidentialVoid {v: Void}
+
+impl CommitEncode for ConfidentialVoid {
+    fn commit_encode<E: io::Write>(self, _e: E) -> usize {0}
+}
+impl From<ConfidentialVoid> for Void {
+    fn from(c: ConfidentialVoid) -> Self {
+        c.v
+    }
+}
+
 impl Conceal for Void {
-    type Confidential = Void;
+    type Confidential = ConfidentialVoid;
 
     fn conceal(&self) -> Self::Confidential {
-        self.clone()
+        ConfidentialVoid {v: self.clone()}
     }
```

Now, the library can be compiled with no error, and the test function can then be called (see example below).

```shell
cargo test -p rgb-core test_void_encoding
```

Explaining the change above: when we replace the `Confidential` type of `Void` with something other than `Void` itself (the new `ConfidentialVoid` type in this case), we break (what I have referred to in one of my previous comments) the "circular condition" imposed by the code bellow.

```rust
    impl<T> CommitEncode for amplify::Holder<T, UsingConceal>
    where
        T: Conceal,
        <T as Conceal>::Confidential: CommitEncode,
    {
        fn commit_encode<E: io::Write>(self, e: E) -> usize {
            self.into_inner().conceal().commit_encode(e)
        }
    }
```

Where for the `CommitEncode` trait to be implemented for `T` (`Void` in our case), `T` must implement the `Conceal` trait (and that part is OK, since `Void` does implement it), and the `Confidential` type defined in the `Conceal` implementation for `T` (in other words, the type returned by the `conceal()` method of `T`) must implement the `CommitEncode` trait. Here is where the problem lies. If we substitute `T` for `Void` we can state the previous condition as: for the `CommitEncode` trait to be implemented for `Void`, `Void` must implement the `CommitEncode` trait, since `Void::conceal()` returns `Void`.

Reply:
@dr-orlovsky and @rajarshimaitra, I just wanted to add that the issue with the implementation of the `CommitEncode` trait for the `Void` type that we have discussed above also applies to other RGB node types, namely `Assignments` and `OwnedState<STATE>`. That is the reason why we do need an explicit implementation of the `CommitEncode` trait for those types.